README 
	Pipeline for 
	Last edited on 10/13/2021 -- GAREEVES

PIPELINE OVERVIEW
	This pipeline is used to accomplish two tasks pertaining to Transposable Element (TE) data integration. The first tasks is to evaluate the level
	of conservation between TEs across species and provide this data for the pipeline detailed in ~/HOME/Motif_Analysis/2_Motif_Coordinates/. The second
	tasks involves counting the amount of overlap between TEs and various sub-groups of ATACseq peaks including a) Concensus ATACseq peaks, B) Diapause
	differential chromatin peaks, and C) control regions down/up stream of chromatin peaks. These overlaps are then broken down by TE class and family
	and cross-compared for enrichment.

HARDWARE SPECIFICATIONS	
	The pipeline is designed to work on a MACOSX device (minimum version BigSir 11.4)	

TEST CASE
	No test data provided (Can be generated using other pipelines)
	

------------------------------------------------------------------------------------------	
DIRECTORY SETUP									|
	./HOME/										| -- The directory for the pipeline
		/Data/									| -- The directory for input and output data
		/Packages/								| -- The directory for storing required packages 
												|
REQUIRED FILES									|
	INPUT DATA									|
		<SPECIES>_overlap.bed					| - A file overlapping TEs with aligned coordinates in African turquoise killifish, generated by REPEATMASKER and BEDTOOLS
		Peak_Master_0.0.txt						| - A file with peak conservation between fish species, generated in ~/HOME/ATACseq_Analysis/3_Conservation/
		Interest_Motif_sort.bed					| - A file with the bed-formated coordinates of all motifs in African turquoise killifish, generated in ~/HOME/Motif_Analysis/2_Motif_Coordinates/
		Nfur_TE.bed								| - A file with bed-formated coordinates of all TEs in African turquoise killifish, generated by REPEATMASKER
		Nfur_True_TE.bed						| - A filtered version of Nfur_TE.bed to correct names of TE families/classes
		Nfur_True_DE.bed						| - A filtered version of diapause differential chromatin in the African turquoise killifish, generated in ~/HOME/ATACseq_Analysis/2_Peak_Calling/
		Nfur_True.bed							| - A filtered list of all chromatin peaks in African turquoise killifish, generated in ~/HOME/ATACseq_Analysis/2_Peak_Calling/
												|
REQUIRED PACKAGES								|
	bedtools									| -- https://github.com/arq5x/bedtools2 (does not allow subdirectories)
		bedtools intersect						|
												|
	stats										|-- https://www.r-project.org/
	MutationalPatterns							|-- https://bioconductor.org/packages/release/bioc/html/MutationalPatterns.html
	ggplot2										|-- https://ggplot2.tidyverse.org/
	circlize									|-- https://cran.r-project.org/web/packages/circlize/index.html
												| 
	SCRIPTS										|
		TE_Motif.py								| -> placed in ~/HOME/
		TE_Dater.py								| -> placed in ~/HOME/
		TE_Bubble_Prep.py						| -> placed in ~/HOME/
		TE_Class_Filter.py						| -> placed in ~/HOME/
		Bubble.R								| -> placed in ~/HOME/
												|
												|
	SCRIPT RUN ORDER							|
												|
		TE_Motif.py								|
		|										|
		V										|
		TE_Dater.py								|
												|
		TE_Bubble_Prep.py						|
		|										|
		V										|
		TE_Class_Filter.py						|
		|										|
		V										|
		Bubble.R								|  
												|	
PRE-RUNNING INSTRUCTIONS						|
	Script updates								|
		TE_Motif.py								| -> Path update
												|	Update the path to home directory on line 13
												| -> Package Path Updates
												|	if paths to all packages are in your .bashrc/.bash_profile, then no update is required
												|	else, adjust the path to package functions on line 14
												| 
												|
		TE_Dater.py								| -> Path update
												|	Update the path to home directory on line 11
												| 
		TE_Bubble_Prep.py						| -> Path update
												|	Update the path to home directory on line 13
												| -> Package Path Updates
												|	if paths to all packages are in your .bashrc/.bash_profile, then no update is required
												|	else, adjust the path to package functions on line 14
												| 
		TE_Class_Filter.py						| -> Path update
												|	Update the path to home directory on line 11
												|										
		Bubble.R								|-> Path update
												|	Update the path to home directory on line 11
												|
												|
												|
------------------------------------------------------------------------------------------
		