README 
	Pipeline For positive selection and metadata integration of regulatory regions
	Last edited on 10/13/2021 -- GAREEVES

PIPELINE OVERVIEW
	This pipeline takes sequences from multiple genome alignments files that overlap with chromatin accessibility peaks of interest
	and uses them to infer the ancestral state of the sequence at various nodes along a phylogenetic tree. Using these inferred sequences,
	we performed a test for positive selection signature at regulatory regions as described in the below listed publication:
	
	 	Jialin Liu, Rebecca R. Viales, Pierre Khoueiry, James P. Reddington, Charles Girardot, 
	 	Eileen E. M. Furlong, Marc Robinson-Rechavi. The hourglass model of evolutionary conservation 
	 	during embryogenesis extends to developmental enhancers with signatures of positive selection. 
	 	Genome Research. July, 2021. doi:10.1101/gr.275212.121
	
	After performing the pipeline as describe, we then integrated metadata from our other analysis including motif, TE, and gene overlaps to
	the sites. We then also used this pipeline to merge our finding at two distinct nodes of ancestral inferance, using the super set to do out
	final manual curation for the paper.

HARDWARE SPECIFICATIONS	
	The pipeline is designed to work on a MACOSX device (minimum version BigSir 11.4)

TEST CASE
	No test data provided (Can be generated using other pipelines)
	

------------------------------------------------------------------------------------------	
DIRECTORY SETUP									|
	./HOME/										| -- The directory for the pipeline
		/Data/									| -- The directory for input and output data
			/Chunk_Seqs/						| -- The directory for 
			/Peak_Seqs_DE/						| -- The directory for
												|
		/Packages/								| -- The directory for storing required pacakages for the pipeline
												|
REQUIRED PACKAGES								|
	Bedtools									|-- https://github.com/arq5x/bedtools2 (does not allow subdirectories)
		bedtools intersect						|
												|
	PAML										|-- http://abacus.gene.ucl.ac.uk/software/paml.html
		baseml									|
												|
REQUIRED FILES									|
	INPUT DATA									|
		Peak<NUMBER>.<NUMBER>.fa				| - fasta-formated maf blocks of sequences from African Turquoise killifish, split from maf generated in ~/HOME/Multi_Alignment_Construction/
		PEAK<NUMBER>.fa							| - fasta-formated sequences of chromatin accessibility peaks in the African turquoise killifish, split from peaks generated in ~/HOME/ATACseq_Analysis/2_Peak_Calling/
		Paralog_List.txt						| - File with a list of paralog pairs, generated by ~/HOME/Paralog_Classification/
		Final_Con_List_peak_Edit_R.txt			| - File containing all metadata pertaining to chromatin accessibility conservation between several fish species, generated in ~/HOME/ATACseq_Analysis/3_Conservation/
		fish4_WGA.maf							| - Multiple, whole genome alignment between several fish species, generated in ~/HOME/Multi_Alignment_Construction/
		Interest_Motif_sort.bed					| - Bed-formatted coordinates for all enriched motifs within diapause-accessible chromatin of the African turquoise killifish, generated in ~/HOME/Motif_Analysis/2_Motif_Coordinates/
		Nfur_Peak_All.bed						| - All accessible chromatin sites in the African turquoise killifish, generated in ~/HOME/ATACseq_Analysis/2_Peak_Calling/
		fish.tree								| - Phylogenetic tree used by PAML/baseml to infer the relationship between species. Formatted as directed in the PAML manual
												|
		Chunk_All_KS.txt						| - The output of positive selection pipeline as described in XXX et al. (Provided for ease of avoiding this step when running)
		Chunk_ALL_PS.txt						| - The output of positive selection pipeline as described in XXX et al. (Provided for ease of avoiding this step when running)
												|
	SCRIPTS										|
		Infer_Ancestor_Chunks.py				| -> placed in ~/HOME/
		PS_Meta_Data.py							| -> placed in ~/HOME/
		PS_Merger.py							| -> placed in ~/HOME/
												|
	SCRIPT RUN ORDER							|
												|
		Infer_Ancestor_Chunks.py				|
		|										|
		V										|
		<POSITIVE SELECTION PIPELINE>			| -- This pipeline is described and credited in XXX et al. 
		|										|
		V										|
		PS_Meta_Data.py							| 
		|										|
		V										| 
		PS_Merger.py 							| 
												|	
PRE-RUNNING INSTRUCTIONS						|
	Script updates								|
		Infer_Ancestor_Chunks.py				| -> Path update
												|	Update the path to home directory on line 14
												| -> Package Path update
												|	if paths to all packages are in your .bashrc/.bash_profile, no update needed
												|	else update path to package function calls on line 15
												| -> baseml input parameters
												|	all baseml parameters can be found and edited on line 19
												|
		PS_Meta_Data.py							| -> Path update
												|	Update the path to home directory on line 13
												| -> Package Path update
												|	if paths to all packages are in your .bashrc/.bash_profile, no update needed
												|	else update path to package function calls on line 14
												| -> Ancestral Run sets
												|	To include/exclude particular ancestral nodes, edit file list on line 15
												| 
		PS_Merger.py							| -> Path update
												|	Update the path to home directory on line 11
												| 
												|*NOTE: for final submission, further manual curating was done on merged list									
												|
												|
------------------------------------------------------------------------------------------
		