README 
	Pipeline for assessment of motif coordinates and conservation metadata
	Last edited on 10/12/2021 -- GAREEVES

PIPELINE OVERVIEW
	This pipeline takes a set of all and detected motifs in the African turquoise killifish and uses various other genomic data
	to add pertinent metadata for each detected instance of a motif. The motifs are cross-referenced with both ATACseq peaks and
	Transposable element (TE) coordinates to determine their percentage overlap with each. This is then extended to include aligned
	regions in other fish species contained in a multiple whole genome alignment of 5 fish. These orthologous site are then evaluated
	to generate binding likelihood scores to determine if the binding site has the conservation level needed to bind the given protein/
	transcription factor in the other species.

HARDWARE SPECIFICATIONS	
	The pipeline (All but MasterNT.sh and Subscripts) is designed to work on a MACOSX device (minimum version BigSir 11.4)
	
	The pipeline (MasterNT.sh and Subscripts) is designed to work on a cluster using SLURM for job submission and a
	86x64 ubuntu/linux architecture.

TEST CASE
	No test data provided (Can be generated using other pipelines)
	

------------------------------------------------------------------------------------------	
DIRECTORY SETUP									|
	./HOME/										| -- The directory for the pipeline
		/Data/									| -- The directory for input and output data
			/DEC/								| -- The directory for updated RAW motif files
			/err/								| -- The output for standard error of MasterNT.sh on SLURM
			/Final/								| -- The directory for finalized motif files
			/Interest_Beds/						| -- The directory for bed-formatted motif coordinates
			/Logos/								| -- The directory for short sequences to generate logo graphs
			/out/								| -- The output for standard out of MasterNT.sh on SLURM
			/Raw/								| -- The directory for raw motif summary files (provided for ease)
		/Packages/								| -- The directory for storing required pacakages for the pipeline
		/Subscripts/							| -- The directory containing scripts for job submission by MasterNT.sh
												|
REQUIRED PACKAGES								|
	Bedtools									|-- https://github.com/arq5x/bedtools2 (does not allow subdirectories)
		bedtools intersect						|
		bedtools sort							|	
												|
REQUIRED FILES									|
	INPUT DATA									|
		AllMotifs_ConsensusPeaks_<species>.txt	| - Files generated by homer containing all identified motifs
		Para_DW.txt								| - File linking paralogs to differential chromatin peaks
		Para_UP.txt								| - File linking paralogs to differential chromatin peaks
		Paralog_List.txt						| - File with a list of paralog pairs, generated by ~/HOME/Paralog_Classification/
		Peak_Master_0.0.txt						| - File with cross-species peak conservation, generated by ~/HOME/ATACseq_Analysis/3_Conservation/
		TE_Conservation.txt						| - File with cross-species TE conservation, generated by ~/HOME/Motif_Analysis/4_Transposons_and_Motifs/
		fish4_WGA.maf							| - Multiple whole genome alignment file, generated by ~/HOME/Multi_Alignment_Construction/
		Master_DE_Dia_up_Peeak.bed				| - File containing list of all differential diapause chromatin peaks, generated ~/HOME/ATACseq_Analysis/2_Peak_Calling/
		master_motif_file.motif					| - File containing all known motif sequences provided by HOMER
		Nfur_Gene_Peak_Overlap.bed				| - File with bed-format overlap between ATACseq peaks and Genes in African turquoise killifish
		Nfur_TE.bed								| - File with bed-format coordinates for all TEs in African turquoise killifish, generated by REPEATMASKER
		TE_DE_DW.bed							| - File with overlap between TEs and differential chromatin, generated by ~/HOME/Motif_Analysis/4_Transposons_and_Motifs/
		TE_DE_UP.bed							| - File with overlap between TEs and differential chromatin, generated by ~/HOME/Motif_Analysis/4_Transposons_and_Motifs/
		TE_Motif_Overlap.bed					| - Overlap between TEs and Motifs in African turquoise killifish, generated by ~/HOME/Motif_Analysis/4_Transposons_and_Motifs/
		<motif>.txt								| - RAW output files from MasterNT.sh run on SLURM system. (provided to avoid need to rerun script on cluster)
												|
	SCRIPTS										|
		I_Motif.py								| -> placed in ~/HOME/
		MasterNT.sh								| -> placed in ~/HOME/
		Motif_Bed_Splitter.py					| -> placed in ~/HOME/
		Motif_Correction.py						| -> placed in ~/HOME/
		Motif_Master_Cleanup.py					| -> placed in ~/HOME/
		Motif_Meta_Summary.py					| -> placed in ~/HOME/
												|
	SUB-SCRIPTS									|
		TN_Submission.sh						| -> placed in ~/Home/Sub_Scripts/
		Trans_Novo.py							| -> placed in ~/Home/Sub_Scripts/ 
												|
	SCRIPT RUN ORDER							|
												|
		I_Motif.py								|
		|										|
		V										|
		Motif_Bed_Splitter.py					|
		|										|
		V										|
		MasterNT.sh (-> NT_submission.sh)		| 
		|						|				|
		V						V				| 
		Motif_Correction.py	   (Trans_Novo.py)	| 
		|										| 
		V										|	
		Motif_Master_Cleanup.py					|	
		|										| 
		V										| 
		Motif_Meta_Summary.py					|	 
												|	
PRE-RUNNING INSTRUCTIONS						|
	Script updates								|
		I_Motif.py								| -> Path update
												|	Update the path to home directory on line 12
												| -> Package Path update
												|	if paths to all packages are in your .bashrc/.bash_profile, no update needed
												|	else update path to package function calls on line 13
												|
		MasterNT.sh								| -> Path update
												|	Update the path to home directory on line 9
												| 
		Motif_Bed_Splitter.py					| -> Path update
												|	Update the path to home directory on line 12
												| 
		Motif_Correction.py						| -> Path update
												|	Update the path to home directory on line 14
												| 
		Motif_Master_Cleanup.py					| -> Path update
												|	Update the path to home directory on line 15
												| 
		Motif_Meta_Summary.py					| -> Path update
												|	Update the path to home directory on line 15
												|
	Sub-Script updates							| 
		TN_Submission.sh						|-> Path update
												|	Update the path to home directory on line 8
												|
		Trans_Novo.py							|-> Path update
												|	Update the path to home directory on line 14
												|										
												|
												|
------------------------------------------------------------------------------------------
		